src/Psl/Type/Internal/ConvertedType.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Type\Internal;
+
+use Closure;
+use Psl\Type;
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+use Psl\Type\TypeInterface;
+use Throwable;
+
+/**
+ * @template I
+ * @template O
+ *
+ * @ara-extends Type\Type<O>
+ *
+ * @extends Type\Type<O>
+ *
+ * @internal
+ */
+final class ConvertedType extends Type\Type
+{
+    /**
+     * @param TypeInterface<I> $from
+     * @param TypeInterface<O> $into
+     * @param (\Closure(I): O) $converter
+     */
+    public function __construct(
+        private TypeInterface $from,
+        private TypeInterface $into,
+        private Closure $converter
+    ) {
+    }
+
+    /**
+     * @throws CoercionException
+     *
+     * @ara-return O
+     *
+     * @return O
+     */
+    public function coerce(mixed $value): mixed
+    {
+        if ($this->into->matches($value)) {
+            return $value;
+        }
+
+        $coercedInput = $this->from->coerce($value);
```

Comment:
should we try to coerce it into the `into` type first? maybe that is possible, even if it doesn't match

Reply:
I'dd say that would be a strange limitation. If you explicitly pass a converter function, it is strange to do some auto-conversion internal first?

For example: if you want to have a string that has a specific format, you could parse it like this now:


```php
$scalarEmailType = Type\converted(
    Type\string(),
    Type\non_empty_string(),
    static function (string $value): string {
         if (!Regex\Matches($value, $emailPattern)) {
               throw new \RuntimeException('The provided email is not valid');
         }
         return $value;
    }
);
```

Reply:
This way, it could solve scenarios like these:
https://github.com/azjezz/psl/pull/404

```php
$customBoolType = Type\converted(
    Type\string(),
    Type\bool(),
    static function (string $value): bool {
        return !in_array($value, ['', 'False', 'No'], true);
    }
);
```